### PR TITLE
UCT: DC error handling

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
@brminich @shssf pls review

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1149/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3080/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
@yosefe pls review as well

File: src/uct/ib/mlx5/ib_mlx5.c [hunk 1]
```diff
@@ -243,7 +243,7 @@ struct mlx5_cqe64* uct_ib_mlx5_check_completion(uct_ib_iface_t *iface,
     case MLX5_CQE_REQ_ERR:
         iface->ops->handle_failure(iface, cqe);
         ++cq->cq_ci;
-        return NULL;
+        return cqe;
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.c]: why is this needed? looks like it makes us add extra branch in poll_tx

File: src/uct/ib/dc/accel/dc_mlx5.c [hunk 2]
```diff
@@ -553,8 +554,22 @@ static void uct_dc_mlx5_iface_progress(void *arg)
...
-    struct mlx5_cqe64 *cqe = arg;
-    uct_ib_mlx5_completion_with_err((void*)cqe, 0);
+    struct mlx5_cqe64   *cqe    = arg;
+    uint32_t            qp_num  = ntohl(cqe->sop_drop_qpn) & UCS_MASK(UCT_IB_QPN_ORDER);
+    uct_dc_mlx5_iface_t *iface  = ucs_derived_of(ib_iface, uct_dc_mlx5_iface_t);
+    uint8_t             dci     = uct_dc_iface_dci_find(&iface->super, qp_num);
+    uct_dc_ep_t         *ep     = iface->super.tx.dcis[dci].ep;
+
+    if (ep) {
+        uct_rc_txqp_purge_outstanding(&iface->super.tx.dcis[dci].txqp,
```
Comment (yosefe) [inline src/uct/ib/dc/accel/dc_mlx5.c]: can extract this code to common function for dc_mlx5 and dc_verbs
also, need to revive the dci since after the failure the failed dci becomes unusable

File: src/uct/ib/dc/verbs/dc_verbs.c [hunk 3]
```diff
@@ -96,7 +96,8 @@ static ucs_status_t uct_dc_verbs_iface_query(uct_iface_h tl_iface, uct_iface_att
                                       UCT_IFACE_FLAG_ATOMIC_CSWAP32|
                                       UCT_IFACE_FLAG_ATOMIC_DEVICE |
                                       UCT_IFACE_FLAG_PENDING|
-                                      UCT_IFACE_FLAG_AM_CB_SYNC|UCT_IFACE_FLAG_CONNECT_TO_IFACE;
+                                      UCT_IFACE_FLAG_AM_CB_SYNC|UCT_IFACE_FLAG_CONNECT_TO_IFACE|
```
Comment (yosefe) [inline src/uct/ib/dc/verbs/dc_verbs.c]: pls add flag on new line

File: src/uct/ib/dc/base/dc_ep.c [hunk 4]
```diff
@@ -70,6 +70,27 @@ void uct_dc_ep_release(uct_dc_ep_t *ep)
...
+    ucs_status_t status;
+
+    if (ep) {
+        uct_rc_txqp_purge_outstanding(&iface->tx.dcis[dci].txqp,
+                                      UCS_ERR_ENDPOINT_TIMEOUT, 0);
+        uct_set_ep_failed(ep_cls, &ep->super.super,
+                          &iface->super.super.super.super);
+        status = uct_dc_iface_dci_reconnect(iface, &iface->tx.dcis[dci].txqp);
+        if (status != UCS_OK) {
+            ucs_fatal("Unsuccessful DC QP reconnect");
```
Comment (miked-mellanox) [inline src/uct/ib/dc/base/dc_ep.c]: maybe print qp_num here

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1168/ for details.

File: src/uct/ib/dc/base/dc_iface.c [hunk 5]
```diff
@@ -132,6 +132,27 @@ static ucs_status_t uct_dc_iface_dci_connect(uct_dc_iface_t *iface, uct_rc_txqp_
...
 
+/* restore dc qp from error to rts state */
+ucs_status_t uct_dc_iface_dci_reconnect(uct_dc_iface_t *iface,
+                                        uct_rc_txqp_t *dci)
+{
+    struct ibv_exp_qp_attr attr;
+
+    memset(&attr, 0, sizeof(attr));
+    attr.qp_state        = IBV_QPS_RESET;
+    attr.pkey_index      = 0;
```
Comment (shamisp) [inline src/uct/ib/dc/base/dc_iface.c]: you already "zero" everything with memset(), not sure what is the purpose of the following lines

Comment (evgeny-leksikov) [inline src/uct/ib/dc/base/dc_iface.c]: Copy-paste issue... fill be fixed

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1169/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3103/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3104/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3108/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
mlx:bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1173/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3109/ for details (Mellanox internal link).

File: src/uct/ib/dc/base/dc_ep.c [hunk 6]
```diff
@@ -70,6 +70,27 @@ void uct_dc_ep_release(uct_dc_ep_t *ep)
...
 }
 
+void uct_dc_ep_set_failed(ucs_class_t *ep_cls, uct_dc_iface_t *iface,
+                          uint32_t qp_num)
+{
+    uint8_t dci = uct_dc_iface_dci_find(iface, qp_num);
+    uct_dc_ep_t *ep = iface->tx.dcis[dci].ep;
+    ucs_status_t status;
+
+    if (ep) {
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_ep.c]: if (!ep) { return; }

File: src/uct/ib/dc/base/dc_iface.c [hunk 7]
```diff
@@ -132,6 +132,23 @@ static ucs_status_t uct_dc_iface_dci_connect(uct_dc_iface_t *iface, uct_rc_txqp_
...
+/* restore dc qp from error to rts state */
+ucs_status_t uct_dc_iface_dci_reconnect(uct_dc_iface_t *iface,
+                                        uct_rc_txqp_t *dci)
+{
+    struct ibv_exp_qp_attr attr;
+
+    memset(&attr, 0, sizeof(attr));
+    attr.qp_state = IBV_QPS_RESET;
+
+    if (ibv_exp_modify_qp(dci->qp, &attr, IBV_EXP_QP_STATE)) {
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.c]: pls take a look on rc_mlx5_ep.c:466:
1. need to reuse code from uct_rc_ep_reset_qp (provavly create common function to accept uct_rc_txqp_t*)
2. for mlx5, need to call uct_ib_mlx5_update_cq_ci() before, and update cq_ci after, the call to reset. this is because the driver can change internal cq->cons_index when modifying qp to reset and need to resync it. pls extract common code for that with rc_mlx5, if possible.

File: src/uct/ib/dc/verbs/dc_verbs.c [hunk 8]
```diff
@@ -603,6 +605,19 @@ ucs_status_t uct_dc_verbs_iface_create_ah(uct_dc_iface_t *dc_iface, uint16_t lid
     return UCS_OK;
 }
 
+static UCS_F_NOINLINE void uct_dc_verbs_handle_failure(uct_ib_iface_t *ib_iface,
+                                                       void *arg)
+{
+    struct ibv_wc   *wc     = arg;
```
Comment (yosefe) [inline src/uct/ib/dc/verbs/dc_verbs.c]: no need for multiple spaces. also no need for UCS_F_NOINLINE since it's called through polymorphic function pointer anyway.

File: src/uct/ib/dc/base/dc_iface.c [hunk 9]
```diff
@@ -134,19 +136,29 @@ static ucs_status_t uct_dc_iface_dci_connect(uct_dc_iface_t *iface, uct_rc_txqp_
...
 ucs_status_t uct_dc_iface_dci_reconnect(uct_dc_iface_t *iface,
-                                        uct_rc_txqp_t *dci)
+                                        uct_rc_txqp_t *txqp,
+                                        ucs_class_t *ep_cls)
 {
-    struct ibv_exp_qp_attr attr;
-
-    memset(&attr, 0, sizeof(attr));
-    attr.qp_state = IBV_QPS_RESET;
+    uct_rc_mlx5_iface_t *mlx5_iface;
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.c]: need to avoid mlx5/verbs code in "base". maybe better to add another virtual function to rc_iface: reset_qp(iface, txqp)?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1178/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3120/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1179/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1203/ for details.

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3151/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3153/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3154/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1213/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1215/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3163/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3164/ for details (Mellanox internal link).

File: src/uct/ib/dc/base/dc_iface.c [hunk 10]
```diff
@@ -132,6 +133,21 @@ static ucs_status_t uct_dc_iface_dci_connect(uct_dc_iface_t *iface, uct_rc_txqp_
     return UCS_OK;
 }
 
+/* restore dc qp from error to rts state */
+ucs_status_t uct_dc_iface_dci_reconnect(uct_dc_iface_t *iface,
```
Comment (evgeny-leksikov) [inline src/uct/ib/dc/base/dc_iface.c]: Just to improve readability and consistency, it will complicate `uct_dc_ep_set_failed` function a bit - one more error check with fatal message.

Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.c]: ok
